crates/ide/src/inlay_hints/closure_captures.rs

```diff
@@ -0,0 +1,193 @@
+                text_edit: None,
+            });
+            range
+        }
+    };
+    acc.push(InlayHint {
+        range: move_kw_range,
+        kind: InlayKind::ClosureCapture,
+        label: InlayHintLabel::from("("),
+        text_edit: None,
+    });
+    let last = captures.len() - 1;
+    for (idx, capture) in captures.into_iter().enumerate() {
+        let local = capture.local();
+        let source = local.primary_source(sema.db);
+
+        // force cache the source file, otherwise sema lookup will potentially panic
+        _ = sema.parse_or_expand(source.file());
+
+        acc.push(InlayHint {
+            range: move_kw_range,
+            kind: InlayKind::ClosureCapture,
+            label: InlayHintLabel::simple(
+                format!(
+                    "{}{}",
+                    match capture.kind() {
+                        hir::CaptureKind::SharedRef => "&",
+                        hir::CaptureKind::UniqueSharedRef => "&unique ",
+                        hir::CaptureKind::MutableRef => "&mut ",
+                        hir::CaptureKind::Move => "",
+                    },
+                    local.name(sema.db)
+                ),
+                None,
+                source.name().and_then(|name| sema.original_range_opt(name.syntax())),
+            ),
+            text_edit: None,
+        });
+
+        if idx != last {
+            acc.push(InlayHint {
+                range: move_kw_range,
+                kind: InlayKind::ClosureCapture,
+                label: InlayHintLabel::simple(", ", None, None),
+                text_edit: None,
+            });
+        }
+    }
+    acc.push(InlayHint {
+        range: move_kw_range,
+        kind: InlayKind::ClosureCapture,
+        label: InlayHintLabel::from(")"),
+        text_edit: None,
+    });
+
+    Some(())
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        inlay_hints::tests::{check_with_config, DISABLED_CONFIG},
+        InlayHintsConfig,
+    };
+
+    #[test]
+    fn all_capture_kinds() {
+        check_with_config(
+            InlayHintsConfig { closure_capture_hints: true, ..DISABLED_CONFIG },
+            r#"
+//- minicore: copy, derive
+
+
+#[derive(Copy, Clone)]
+struct Copy;
+
+struct NonCopy;
+
+fn main() {
+    let foo = Copy;
+    let bar = NonCopy;
+    let mut baz = NonCopy;
+    let qux = &mut NonCopy;
+    || {
+// ^ move
+// ^ (
+// ^ &foo
+// ^ ,
+// ^ bar
+// ^ ,
+// ^ baz
+// ^ ,
+// ^ qux
+// ^ )
+        foo;
+        bar;
+        baz;
+        qux;
+    };
+    || {
+// ^ move
+// ^ (
+// ^ &foo
+// ^ ,
+// ^ &bar
+// ^ ,
+// ^ &baz
+// ^ ,
+// ^ &qux
+// ^ )
+        &foo;
+        &bar;
+        &baz;
+        &qux;
+    };
+    || {
+// ^ move
+// ^ (
+// ^ &mut baz
+// ^ )
+        &mut baz;
+    };
+    // FIXME: &mut qux should be &unique qux
+    || {
+// ^ move
+// ^ (
+// ^ &mut baz
+// ^ ,
+// ^ &mut qux
+// ^ )
+        baz = NonCopy;
+        *qux = NonCopy;
+    };
```

Comment:
`qux` is not mutable, but we are taking a mutable reference to the `*qux`, not `qux` itself. For example this code:
```Rust
    let mut t1 = 2;
    let mut t2 = 5;
    let mut x = &mut t1;
    let mut y = || {
        //&x;
        *x = 3;
    };
    x = &mut t2;
    y();
```
compiles since `y` is capturing `*x` by mutable borrow, so `x` itself is free to mutate. But if the `//&x;` line becomes uncommented, `y` will capture the whole `x` by unique immutable borrow, and we will get compile error.

Reply:
I think it will becomes less confusing if we render the whole place.

Reply:
I think my confusion comes from thinking of old capture logic? Also the rust reference I suppose is outdated then. You are certainly right that we should be rendering the place we capture, as with this mere field captures will render incorrectly already I believe.

Reply:
> I think my confusion comes from thinking of old capture logic? Also the rust reference I suppose is outdated then

Yes the rust reference section for closure is pre edition-2021, and is currently in a misleading state.

Reply:
Turns out there is a PR for this already so I'll dig into that for now https://github.com/rust-lang/reference/pull/1059

Reply:
The new writing still refers to this as a unique immutable borrow fwiw https://github.com/sexxi-goose/reference/blob/closure-doc/src/types/closure.md#unique-immutable-borrows-in-captures